### PR TITLE
[WB-5095] Ignore wandb.log() calls from forked processes

### PR DESCRIPTION
**Pull request by raubitsj**
https://wandb.atlassian.net/browse/WB-5095

Description
-----------

wandb does not support logging from other processes (right now).  All calls to methods should be done from the same process that called wandb.init()

Changes:
1. warning and ignore calls to wandb.log() from the wrong process
2. move exceptions from wandb.errors.error.Error to wandb.errors.Error
3. add multiprocessing tests
4. add strict mode to make some warnings turn into failures (useful for finding incompatible sdk usages)

If you call wandb.log() from an invalid process context you get:
```
wandb: WARNING log() ignored (called from pid=46691, init called from pid=46641). See: https://docs.wandb.ai/library/init#multiprocess
```

There is work in progress to make wandb multiprocess safe.

Known areas that are not safe are:
- wandb.log() we keep track of state in the calling process,  this logic is being moved to the backend process.

TODO:
- [x] Add test
- [x] add strict mode
- [x] add specific exception
- [x] get coverage to 100%
- [x] debug windows test failure

Wishlist / Future:
- add check in interface to cover all cases?  error handling is hard at this layer so will do it later.
- telemetry to track when this happens. this requires some parts of the wandb library to be multiprocess safe (which it probably is)
- other methods need to be handled, possibly any user of wandb object.
- console is really messed up with forking: https://wandb.atlassian.net/browse/WB-5096

Testing
-------

unittests, regression tests

### CODE REVIEW

**Review by vanpelt — APPROVED** [review]
Nice

File: tests/test_mp_full.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+
+def train(add_val):
+    time.sleep(1)
+    wandb.log(dict(mystep=1, val=2 + add_val))
+    wandb.log(dict(mystep=2, val=8 + add_val))
+    wandb.log(dict(mystep=3, val=3 + add_val))
+    wandb.log(dict(val2=4 + add_val))
+    wandb.log(dict(val2=1 + add_val))
+    time.sleep(1)
```
Comment (KyleGoyette) [inline tests/test_mp_full.py:21]: why do we need these sleeps?

File: tests/test_mp_full.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+def test_multiproc_default(live_mock_server, parse_ctx):
+    run = wandb.init()
+    train(0)
+    run.finish()
+
+    ctx_util = parse_ctx(live_mock_server.get_ctx())
+
+    summary = ctx_util.summary
+    s = {k: v for k, v in dict(summary).items() if not k.startswith("_")}
+    assert dict(val=3, val2=1, mystep=3) == s
```
Comment (KyleGoyette) [inline tests/test_mp_full.py:33]: should we convert this check into a function? that way we can just call `check_summary` whenever we want to check the summary of a test using the train function?

**Review by KyleGoyette — COMMENTED** [review]
Changes look good, I suggest pulling the checks for testing summaries into a function to DRY it out. But not critical.

File: tests/test_mp_full.py (hunk 1 above)
Comment (raubitsj) [reply tests/test_mp_full.py:21]: not needed, will remove in next. set of mp tests

File: tests/test_mp_full.py (hunk 2 above)
Comment (raubitsj) [reply tests/test_mp_full.py:33]: agreed, will do in next set of mp test changes.

**Pull request merged**